### PR TITLE
wasm: accurately detect clang when building v8.

### PR DESCRIPTION
**Pull request by jpeach**
Commit Message:

Rather than only assuming that compilers with the string "clang" in
their name are clang, check the built-in preprocessor symbols to ensure
that detection of the compiler type is accurate. This enables the wee8
build to configure itself properly so that compiler-dependent feature
implementations (e.g. pointer compression) build correctly.
    
Additional Description: None
Risk Level: This would affect cross-compilation since the detection is done with the host toolchain. Not sure whether cross-compiling is supported.
Testing: CI build checks
Docs Changes: None
Release Notes: None
Platform Specific Features: None

### CODE REVIEW

File: bazel/external/wee8.genrule_cmd [hunk 1]
```diff
@@ -30,16 +55,23 @@ rm -rf out/wee8
...
+case $$(detect::compiler) in
+clang)
+  export IS_CLANG=true
+  export CXXFLAGS="$${CXXFLAGS} -Wno-implicit-int-float-conversion -Wno-builtin-assume-aligned-alignment -Wno-final-dtor-non-final-class -Wno-invalid-constexpr"
+  ;;
+*)
+  export IS_CLANG=false
+  ;;
+esac
+
```
Comment (PiotrSikora) [inline bazel/external/wee8.genrule_cmd:74]: Unless I'm missing something obvious, this whole PR could be reduced to:
```
if ${CXX} --version 2>/dev/null | grep clang >/dev/null; then
    export IS_CLANG=true
    export CXXFLAGS=...
else
    export IS_CLANG=false
fi
```

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

**Comment by jpeach** [comment]
> On Jun 16, 2021, at 4:07 PM, Piotr Sikora ***@***.***> wrote:
> 
> ﻿
> @PiotrSikora requested changes on this pull request.
> 
> In bazel/external/wee8.genrule_cmd:
> 
> >    export CC=$${CC:-gcc}
>    export CXX=$${CXX:-g++}
>  fi
>  
> +case $$(detect::compiler) in
> +clang)
> +  export IS_CLANG=true
> +  export CXXFLAGS="$${CXXFLAGS} -Wno-implicit-int-float-conversion -Wno-builtin-assume-aligned-alignment -Wno-final-dtor-non-final-class -Wno-invalid-constexpr"
> +  ;;
> +*)
> +  export IS_CLANG=false
> +  ;;
> +esac
> +
> Unless I'm missing something obvious, this whole PR could be reduced to:
> 
> if ${CXX} --version 2>/dev/null | grep clang >/dev/null; then
>     export IS_CLANG=true
>     export CXXFLAGS=...
> else
>     export IS_CLANG=false
> fi
Yeh in practice that ought to give the same result though I feel like testing the compiler is better in principle :)

File: bazel/external/wee8.genrule_cmd [hunk 2]
```diff
@@ -7,6 +7,18 @@ ARCH="$$(uname -m)"
...
+detect::compiler() {
+  if $${CXX} --version 2>&1 | grep -q clang ; then
+    echo clang
+  fi
+
+  if $${CXX} --version 2>&1 | grep -q gcc ; then
+    echo gcc
+  fi
+
+  echo unknown
```
Comment (PiotrSikora) [inline bazel/external/wee8.genrule_cmd:19]: There is no early return, so I think that `unknown` is always emitted, hence the build failure.

File: bazel/external/wee8.genrule_cmd [hunk 3]
```diff
@@ -30,16 +42,23 @@ rm -rf out/wee8
...
   export CC=$${CC:-clang}
   export CXX=$${CXX:-clang++}
-  export CXXFLAGS="$${CXXFLAGS} -Wno-implicit-int-float-conversion -Wno-builtin-assume-aligned-alignment -Wno-final-dtor-non-final-class -Wno-invalid-constexpr"
 else
-  export IS_CLANG=false
   export CC=$${CC:-gcc}
   export CXX=$${CXX:-g++}
 fi
 
+case $$(detect::compiler) in
```
Comment (PiotrSikora) [inline bazel/external/wee8.genrule_cmd:52]: Wouldn't it be easier to simply inline `if $${CXX} --version 2>&1 | grep -q clang ; then` here?

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: bazel/external/wee8.genrule_cmd (hunk 2 above)
Comment (jpeach) [reply bazel/external/wee8.genrule_cmd:19]: Ugh, what was I thinking :-/

Comment (PiotrSikora) [reply bazel/external/wee8.genrule_cmd:19]: Ping?

**Review by PiotrSikora — APPROVED** [review]
Thanks!

**Pull request merged**